portfolio/src/main/java/com/google/sps/servlets/MarkerServlet.java

```diff
@@ -123,4 +149,79 @@ public Location getUserLocation(HttpServletRequest request){
     Location userLocation = new Location(lat, lng);
     return userLocation;
   }
+
+  private boolean validateMarker(double reportLat, double reportLng, String reportDate,
+   String reportTime, String reportCrime) {
+    Query query = new Query(ENTITY_TITLE);
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      double lat = (double) entity.getProperty("lat");
+      double lng = (double) entity.getProperty("lng");
+      String crime = (String) entity.getProperty("crimeType");
+      String date = (String) entity.getProperty("date");
+      String time = (String) entity.getProperty("time");
+
+      if (sameLocation(lat, lng, reportLat, reportLng) && sameDate(reportDate, date) 
+          && sameTime(reportTime, time) && sameCrime(reportCrime, crime)) 
+            return false;
+    }
+    return true;
+  }
+
+  private boolean sameLocation(double markerLat, double markerLng, double reportLat, double reportLng) {
+    double R = 6371.0710; // Radius of the Earth in kilometers
```

Comment:
you don't need this calculation. the imported method SloppyMath.haversinMeters does this for you. Just do `return haversinMeters(markerLat, markerLgn, reportLag, reportLng) < 0.01;`

Reply:
Done!